ark/spLDA/processing.py

```diff
@@ -184,3 +187,121 @@ def create_difference_matrices(cell_table, features, training=True, inference=Tr
+        - A scaled estimate of the standard error of the expected log
+        within-cluster sum of squares.
+
+    """
+    # Calculate the range of each feature column
+    mins, maxs = features.apply(min, axis=0), features.apply(max, axis=0)
+    n, p = features.shape
+    w_kb = []
+    # Create bootstrapped reference data
+    boots = np.zeros(shape=(n * num_boots, p))
+    boot_ind = np.repeat(np.array(range(num_boots)), n)
+    np.random.shuffle(boot_ind)
+    for i in range(p):
+        boots[:, i] = np.random.uniform(low=mins[i], high=maxs[i], size=n * num_boots)
+    # Cluster each bootstrapped sample to get the inertia
+    for b in range(num_boots):
+        boot_array = pd.DataFrame(boots[boot_ind == b, :])
+        boot_clust = KMeans(n_clusters=k).fit(boot_array)
+        w_kb.append(boot_clust.inertia_)
+    # Gap statistic and standard error
+    gap = np.log(w_kb).mean() - np.log(clust_inertia)
+    s = np.log(w_kb).std() * np.sqrt(1 + 1 / num_boots)
+    return gap, s
+
+
+def compute_topic_eda(features, topics, num_boots=25):
+    """Computes various metrics for k-means clustering models to help determine an
+    appropriate number of topics for use in spatial-LDA analysis.
+
+    Args:
+        features (pandas.DataFrame):
+            A DataFrame of featurized cellular neighborhoods.  Specifically, this is one of the
+            outputs of :func:`~ark.spLDA.processing.featurize_cell_table`.
+        topics (list):
+            A list of integers corresponding to the different number of possible topics to
+            investigate.
+        num_boots (int):
+            The number of bootstrap samples to use when calculating the Gap-statistic.
+
+    Returns:
+        dict:
+
+        - A dictionary of dictionaries containing the corresponding metrics for each topic value
+        provided.
+
+    """
+    # Check inputs
+    if num_boots < 25:
+        raise ValueError("Number of bootstrap samples must be at least 25")
+    if min(topics) <= 2:
+        raise ValueError("Number of topics must be greater than 2")
+    if max(topics) >= features.shape[0] - 1:
+        raise ValueError("Number of topics must be less than {}".format(features.shape[0] - 1))
+
+    inertias, silhouettes, gap_stats, gap_sds, pct_vars = {}, {}, {}, {}, {}
+
+    # Compute the total sum of squared pairwise distances between all observations
+    total_ss = np.sum(pdist(features) ** 2) / features.shape[0]
+    for k in topics:
+        cluster_fit = KMeans(n_clusters=k).fit(features)
+        inertias[k] = cluster_fit.inertia_
+        silhouettes[k] = silhouette_score(features, cluster_fit.labels_, metric='euclidean')
+        gap_stats[k], gap_sds[k] = gap_stat(features, k, cluster_fit.inertia_, num_boots)
+        pct_vars[k] = (total_ss - cluster_fit.inertia_) / total_ss
+
+    stats = {"inertia": inertias, "silhouette": silhouettes,
+             "gap_stat": gap_stats, "gap_sds": gap_sds, "percent_var_exp": pct_vars}
+    return stats
```

Comment:
Is there a particular benefit to doing it this way?

Reply:
the problem with `inertias = silhouettes = gap_stats = gap_sds = pct_vars = {}` is that they will all be the same dictionary.  Try adding a key to one of them and it'll appear in all of them.  The benefit of this method is that adding or removing stats in the future will be easier, since it's just adding/removing a member to a list, as opposed to initializing a new variable.

Reply:
Damn you python and you're crazy variable scope and pointers. 💪  FYI this is me shaking my fist at python, not flexing.